sdk/nodejs/cmd/run/run.ts

```diff
@@ -53,14 +64,38 @@ function projectRootFromProgramPath(program: string): string {
     }
 }
 
+async function npmPackageRootFromProgramPath(programPath: string): Promise<string> {
+    // pkg-dir is an ESM module which we use to find the location of package.json
+    // Because it's an ESM module, we cannot import it directly.
+    const { packageDirectory } = await dynamicImport("pkg-dir");
+    // Check if programPath is a directory. If not, then we
+    // look at it's parent dir for the package root.
+    let isDirectory = false;
+    if(fs.existsSync(programPath)) {
+        const fileStat = await fspromises.lstat(programPath);
+        isDirectory = fileStat.isDirectory();
+    }
+    const programDirectory = isDirectory ? programPath : path.dirname(programPath);
+    const pkgDir = await packageDirectory({
+        cwd: programDirectory,
+    });
+    if (pkgDir === undefined) {
+        log.warn("Could not find a package.json file for the program. Using the pulumi program directory as the project root.");
+        return programDirectory;
+    }
+    return pkgDir;
+}
+
+
 function packageObjectFromProjectRoot(projectRoot: string): Record<string, any> {
+    let packageObject = {};
     try {
         const packageJson = path.join(projectRoot, "package.json");
-        return require(packageJson);
-    } catch {
+        packageObject = require(packageJson);
+    } finally {
         // This is all best-effort so if we can't load the package.json file, that's
         // fine.
-        return {};
+        return packageObject;
```

Comment:
I'm not sure I understand this change?
The semantics of return from a `finally` are non-obvious to me, but if I'm reading this correctly, this will behave as follows:

- if there is no error, return the result of `require(packageJson)`
- if there is an error, it will return `{}` because `packageObject` is empty

If that's the behavior this is aiming for, the old version has the same behavior, and is more obvious, IMO.

Reply:
I'll revert this change. I felt it was a style improvement to have a single return in this function. Your description of the behavior is correct; it's a refactor only. 👍🏻